### PR TITLE
Make sure note and lease can't be set to an empty byte array.

### PR DESCRIPTION
**Pull request by winder**
It was possible to put the note/lease fields into an invalid state, added some additional checks.

### CODE REVIEW

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java [hunk 1]
```diff
@@ -745,6 +745,12 @@ public static Transaction createAssetTransferTransaction(// AssetTransaction
         return tx;
     }
 
+    private void setNote(byte[] note) {
+        if (note != null && note.length != 0) {
```
Comment (algonautshant) [inline src/main/java/com/algorand/algosdk/transaction/Transaction.java]: What is a valid empty note field? How can one reset the note field to empty?

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java [hunk 2]
```diff
@@ -655,4 +655,26 @@ public void testTransactionWithLease() throws Exception {
         assertThat(encodedOutBytes).isEqualTo(goldenString);
         TestUtil.serializeDeserializeCheck(stx);
     }
+
+    @Test
+    public void EmptyByteArraysShouldBeRejected() throws Exception {
```
Comment (egieseke) [inline src/test/java/com/algorand/algosdk/transaction/TestTransaction.java]: Not sure that the semantics of empty string and null byte array are the same.

**Comment by rotemh** [comment]
I believe we should allow these two to be empty. We just need to make sure they get serialize properly.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java (hunk 1 above)
Comment (winder) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java]: In order for the note field to be omitted, it needs to be `null` in the Transaction object. We were previously checking for null, but allowing it to be an array of length zero.

**Comment by winder** [comment]
@rotemh it has always been valid to have them be empty, the problem was that according to the serialization algorithm "empty" means equal to the default value, which is set to null. So if someone set it to an array of size 0, it would not be considered empty.

File: src/test/java/com/algorand/algosdk/transaction/TestTransaction.java (hunk 2 above)
Comment (winder) [reply src/test/java/com/algorand/algosdk/transaction/TestTransaction.java]: These fields don't accept a string, so "empty string" doesn't apply to the underlying problem.

File: src/main/java/com/algorand/algosdk/transaction/Transaction.java (hunk 1 above)
Comment (winder) [reply src/main/java/com/algorand/algosdk/transaction/Transaction.java]: You shouldn't reset the note field, which is why I made this method private. Unfortunately all of the fields are public so it isn't strictly immutable.

Other than the groupID and the fee, I would prefer that the transaction object be immutable.

**Comment by rotemh** [comment]
so in order to not set the `note` or `lease` field one simply should use the `setNote` `setLease`?

**Comment by winder** [comment]
For lease, yes simply skip it. For note, all of the constructors have a note field, you would set it to null (or `new note[]{}`).

**Pull request merged**